### PR TITLE
Checks whether path exists before adding it to the paths dictionary

### PR DESCRIPTION
**Pull request by irvinesunday**
Fixes https://github.com/microsoft/OpenAPI.NET.OData/issues/343

### CODE REVIEW

File: src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs [hunk 1]
```diff
@@ -270,5 +271,131 @@ internal static void AddCustomAttributesToExtensions(this IDictionary<string, IO
...
             return false;
         }
+
+
+        /// <summary>
+        /// Gets the entity type of the target <paramref name="segment"/>.
+        /// </summary>
+        /// <param name="segment">The target <see cref="ODataSegment"/></param>
+        /// <returns>The entity type of the target <paramref name="segment"/></returns>
+        internal static IEdmEntityType EntityTypeFromPathSegment(this ODataSegment segment)
```
Comment (andrueastman) [inline src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs:281]:
```suggestion
        private static IEdmEntityType EntityTypeFromPathSegment(this ODataSegment segment)
```

File: src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs [hunk 2]
```diff
@@ -270,5 +271,132 @@ internal static void AddCustomAttributesToExtensions(this IDictionary<string, IO
...
+                }                    
+            }
+            else if (segment is ODataKeySegment keySegment)
+            {
+                return keySegment.EntityType;
+            }
+            else if (segment is ODataOperationSegment)
+            {
+                return segment.EntityTypeFromOperationSegment();
+            }
```
Comment (andrueastman) [inline src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs:307]: A switch statement would be nice to use here.
```suggestion
            switch (segment)
            {
                case ODataNavigationPropertySegment navPropSegment:
                    return navPropSegment.EntityType;
                case ODataNavigationSourceSegment navSourceSegment when navSourceSegment.NavigationSource is IEdmEntitySet entitySet:
                    return entitySet.EntityType();
                case ODataNavigationSourceSegment navSourceSegment when navSourceSegment.NavigationSource is IEdmSingleton singleton:
                    return singleton.EntityType();
                case ODataKeySegment keySegment:
                    return keySegment.EntityType;
                case ODataOperationSegment:
                    return segment.EntityTypeFromOperationSegment();
                default:
                    return null;
            }
```

File: src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs [hunk 3]
```diff
@@ -270,5 +271,132 @@ internal static void AddCustomAttributesToExtensions(this IDictionary<string, IO
...
+                if (path.Count < secondLastSegmentIndex)
+                {
+                    Debug.WriteLine($"Invalid path. Operation not bound to any entity. Path: {pathName}");
+                    return false;
+                }
+
+                ODataSegment lastSecondSegment = path.Segments.ElementAt(path.Count - secondLastSegmentIndex);
+                IEdmEntityType boundEntityType = lastSecondSegment.EntityTypeFromPathSegment();
+
+                IEdmEntityType operationEntityType = lastSegment.EntityTypeFromOperationSegment();
```
Comment (andrueastman) [inline src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs:379]: As these two functions can return null, we should probably check for null returns before using the returned values later on.

File: src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs (hunk 1 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs:281]: This extension method will def. be reused outside this class.

File: src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs [hunk 4]
```diff
@@ -270,5 +271,123 @@ internal static void AddCustomAttributesToExtensions(this IDictionary<string, IO
...
+        /// </summary>
+        /// <param name="segment">The target <see cref="ODataOperationSegment"/>.</param>
+        /// <returns>The entity type of the target <paramref name="segment"/>.</returns>
+        private static IEdmEntityType EntityTypeFromOperationSegment(this ODataSegment segment)
+        {
+            CheckArgumentNull(segment, nameof(segment));
+
+            if (segment is ODataOperationSegment operationSegment)
+            {
+                IEdmOperationParameter bindingParameter = operationSegment.Operation.Parameters.First();
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs:312]:
```suggestion
            if (segment is ODataOperationSegment operationSegment &&
            operationSegment.Operation.Parameters.FirstOrDefault() is IEdmOperationParameter bindingParameter)
            {
```

File: src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs [hunk 5]
```diff
@@ -270,5 +271,123 @@ internal static void AddCustomAttributesToExtensions(this IDictionary<string, IO
...
+                if (path.LastSegment is not ODataOperationSegment lastSegment)
+                {
+                    Debug.WriteLine("Duplicate path: " + pathName);
+                    return false;
+                }
+
+                int secondLastSegmentIndex = 2;
+                if (path.Count < secondLastSegmentIndex)
+                {
+                    Debug.WriteLine($"Invalid path. Operation not bound to any entity. Path: {pathName}");
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs:361]: we could have the same issue with an operation bound to a collection, and have a collection at the root. E.g. have one activate method on directory object collections, and and another activate on users collection. All of a sudden /users/activate becomes a problem.

**Review by baywet — CHANGES_REQUESTED** [review]

File: src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs (hunk 5 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs:361]: _/users/activate_ will not fall within that conditional statement as the no. of segments will be equal to 2

File: src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs (hunk 3 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs]: Resolved

**Review by baywet — DISMISSED** [review]

**Review by baywet — APPROVED** [review]

**Pull request merged**